Feat/tina init

logan-anderson opened the pull request:
- [x] Installs tinacms, styled-components, tina-graphql-gateway and tina-graphql-gateway-cli
- [x] Creates a `.tina/schema.ts` file with a blog collection
- [x] Creates a `content/blog` folder with one item `helloWorld.md`
- [x] Creates a `page/blog/[slug].ts` file with all of the Tina pieces wrapped up in one file
  -  getStaticPaths - runs the query to provide slug variables
  -  getStaticProps - runs the query for a single blog post
  - A default page component export is wrapped by `TinaWrapper`, like the component we have in the starter.
 - [ ] ~~Bonus: should we export that component from one of our packages instead to make it more concise?~~

jeffsee55 commented:
> Bonus: should we export that component from one of our packages instead to make it more concise?

Lets hold off on that for this PR, probably need a bit more discussion

DirtyF commented on packages/tina-graphql-gateway-cli/src/cmds/baseCmds.ts:
```diff
@@ -70,6 +72,12 @@ export const baseCmds: Command[] = [
   },
+  {
+    command: INIT,
+    description: 'Add Tinacms to an exsisting project',
```
```suggestion
    description: 'Add TinaCMS to an existing project',
```

DirtyF commented on packages/tina-graphql-gateway-cli/src/cmds/init/index.ts:
```diff
@@ -0,0 +1,66 @@
+export async function initTina(ctx: any, next: () => void, options) {
+  console.log(
+    successText(
+      'Welcome to tina Init process sit back and relax while we set up tina for you'
```
```suggestion
      'Sit back and relax while we set up Tina for you…'
```

DirtyF commented on packages/tina-graphql-gateway-cli/src/cmds/init/index.ts:
```diff
@@ -46,21 +49,66 @@ export async function installDeps(ctx: any, next: () => void, options) {
+  const baseDir = process.cwd()
+
+  console.log(`
+What just Happened?
```
```suggestion
What just Happened?
```

DirtyF commented on packages/tina-graphql-gateway-cli/src/cmds/init/index.ts:
```diff
@@ -101,10 +102,17 @@ What just Happened?
+  )} start your dev server with 'next dev' and go to http://locahost:3000/demo/blog/HelloWorld.md to check it out"
+
+
+For more Information visit our docs and checkout our getting starter guide
```
```suggestion
For more information visit our docs and check out our getting started guide
```

DirtyF commented on packages/tina-graphql-gateway-cli/src/cmds/init/index.ts:
```diff
@@ -101,10 +102,17 @@ What just Happened?
+
+For more Information visit our docs and checkout our getting starter guide
+
+Dos: https://tina.io/docs/tina-cloud/
```
```suggestion
Docs: https://tina.io/docs/tina-cloud/
```

jeffsee55 commented:
@logan-anderson how can we test this? Is it easiest to just merge it and deploy? We could put this under a canary release

logan-anderson commented:
> @logan-anderson how can we test this? Is it easiest to just merge it and deploy? We could put this under a canary release

@jeffsee55 Yeah lets do that

jeffsee55 commented on packages/tina-graphql-gateway-cli/src/cmds/init/setup-files/index.ts:
```diff
@@ -0,0 +1,257 @@
+        __typename
+        ... on Article_Doc_Data {
+          title
+          author {
```
To keep it simple should that autogenerated `schema.ts` _not_ included an author? We can just remove that entirely to keep things simple

jamespohalloran commented:
Just tested this out on `npx create-next-app --example with-typescript next-demo-app`

This is really awesome! I think this will lead to a huge improvement in conversions. 

I took down a few notes as I went through it:
- npx tina init took a few minutes inially. I wonder if there's a few dependencies we can trim out to speed this up. 

- During the following step, I wasn't sure if I was supposed to run `yarn add tinacms styled-components tina-graphql-gateway tina-graphql-gateway-cli` myself. Maybe this is a matter of formatting, or maybe we don't need to show that bolded line at all (if possible).
<img width="588" alt="Screen Shot 2021-06-11 at 1 59 38 PM" src="https://user-images.githubusercontent.com/3323181/121727364-31a07e80-cac2-11eb-920f-25198c1aae3f.png">
<img width="1281" alt="Screen Shot 2021-06-11 at 2 02 47 PM" src="https://user-images.githubusercontent.com/3323181/121727557-71676600-cac2-11eb-8fd0-c0825741d847.png">
- With the red text “Collection 'posts', path 'content/posts' not found, check that path exists on your filesystem”: it's hard to know if this is an actual error that I need to act on.
- The yellow “please add the following scripts to your package.json” text is hard to see on light-themed terminal
- There's a lot to take in from this initial body of text. I wonder if we should split the Tina Cloud setup into a separate CLII command, or maybe some extra formatting would help.
- “Start your dev server with next dev” -> this doesn’t work, because I’m not running the graphql server. 
Should this be: `Start your dev server with yarn dev or npm run dev`?
- When running `yarn dev` I got this error (see below). I had to remove the author stuff from the page's query, as it's not defined in the schema
<img width="1066" alt="Screen Shot 2021-06-11 at 2 10 58 PM" src="https://user-images.githubusercontent.com/3323181/121727903-e9ce2700-cac2-11eb-8db8-7d1fea5fc03b.png">
- I'm interested to see how users feel about always being in edit-mode on this page. It might not be a problem! But we may want to have a decidated page of our docs that talks about how this usually works.

Apart from the few actual errors I ran into, I don't think any of this needs to block its initial rollout! Great job!!

jeffsee55 commented:
Thanks for going through this @jamespohalloran - great points

> With the red text “Collection 'posts', path 'content/posts' not found, check that path exists on your filesystem”: it's hard to know if this is an actual error that I need to act on.

@logan-anderson I believe that's just because we let the default behavior step in when there isn't a `schema.ts`, but can we intentionally build it _first_ so we don't get that message since we're going to be building it anyway.

> There's a lot to take in from this initial body of text. I wonder if we should split the Tina Cloud setup into a separate CLII command, or maybe some extra formatting would help.

Yeah that may make sense, perhaps the part about configuring the .env and stuff could be done as another command.

DirtyF commented:
> There's a lot to take in from this initial body of text

+1 to point to the next CLI command to  make it a bit less overwhelming, as long as we inform the user about the next steps to take and don't leave them hanging.

logan-anderson commented:
@jamespohalloran I fixed all of those issues thanks for the comments! 

> npx tina init took a few minutes inially. I wonder if there's a few dependencies we can trim out to speed this up.

Moved some obvious ones into dev dependencies like typescript, tsup and some stuff with '@types/' prefix

Ran ` npx depcheck`
And it looks like we can remove a lot more
```
npx: installed 147 in 6.923s
Unused dependencies
* axios
* clear
* cookie-parser
* crypto
* esm
* express-graphql
* figlet
* generator-code
* http-proxy
* inquirer
* isomorphic-unfetch
* jsonwebtoken
* jwks-rsa
* listr
* ms
* open
* ora
* path
* simple-git
* tina-graphql-helpers
* yaml
* yo
```

@jeffsee55 Anything on this list that should not be removed?
